File: inst/modules/sources/tracking/bayesianTracking.R
# Bayesian Tracking
BayesianTracking <- R6::R6Class(
  "BayesianTracking",
  inherit = Tracking,
  
  private = list(
  ),
  
  public = list(
    # function name
    funName = function() {
      paste(
        super$funName(),
        "bayesianTracking",
        sep = cecelia:::CCID_CLASS_SEP
      )
    },
    
    # run
    run = function() {
      # reset image information
      self$resetImageInfo()
      
      self$initLog()
      self$writeLog("Start Bayesian Tracking")
      
      # get object
      cciaObj <- self$cciaTaskObject()
      
      # get visibility
      filterVisibilities <- self$funParamVisibilities("filters", onlyVisible = TRUE)
      
      filters <- self$funParams()$filters
      
      if (!is.null(filterVisibilities))
        filters <- filters[names(filterVisibilities)]
      
      # prepare params
      params <- list(
        taskDir = self$envParams()$dirs$task,
        imRes = cciaObj$omeXMLPixelRes(),
        maxSearchRadius = self$funParams()$maxSearchRadius,
        maxLost = self$funParams()$maxLost,
        trackBranching = self$funParams()$trackBranching,
        minTimepoints = self$funParams()$minTimepoints,
        accuracy = self$funParams()$accuracy,
        probToAssign = self$funParams()$probToAssign,
        noiseInital = self$funParams()$noiseInital,
        noiseProcessing = self$funParams()$noiseProcessing,
        noiseMeasurements = self$funParams()$noiseMeasurements,
        distThresh = self$funParams()$distThresh,
        timeThresh = self$funParams()$timeThresh,
        segmentationMissRate = self$funParams()$segmentationMissRate,
        lambdaLink = self$funParams()$lambdaLink,
        lambdaBranch = self$funParams()$lambdaBranch,
        lambdaTime = self$funParams()$lambdaTime,
        lambdaDist = self$funParams()$lambdaDist,
        thetaTime = self$funParams()$thetaTime,
        thetaDist = self$funParams()$thetaDist,
        filters = filters
      )

      # call python
      if (self$funParams()$calcTrackingStatsOnly == FALSE)
        self$pyScript("bayesian_tracking", params)
      
      if (self$funParams()$calcTrackingStats == TRUE) {
        self$writeLog("Calculate tracking stats")
        self$writeLog(self$funParams()$valueName)
        
        # get tracks
        tracks <- list()
        tracks[[self$funParams()$valueName]] <- cciaObj$tracks(self$funParams()$valueName)
        
        # filter tracks on user defined thresholds
        removeTrackIDs <- c()
        
        if (self$funParams()$maxMeanTurningAngle > 0) {
          # get mean angle for tracks and filter
          meanAngleMeasure <- tracks.measure.fun(
            tracks, celltrackR::meanTurningAngle, "meanTurningAngle",
            idcol = "cell_type")
          
          # get filtered IDs
          removeTrackIDs <- meanAngleMeasure[meanTurningAngle > self$funParams()$maxMeanTurningAngle]$track_id
        }
        
        if (self$funParams()$minDisplacement > 0) {
          # get mean angle for tracks and filter
          displacementMeasure <- tracks.measure.fun(
            tracks, celltrackR::displacement, "displacement",
            idcol = "cell_type")
          
          # get filtered IDs
          removeTrackIDs <- c(
            removeTrackIDs,
            displacementMeasure[displacement < self$funParams()$minDisplacement]$track_id
          )
        }
        
        # filter tracks
        if (length(removeTrackIDs) > 0) {
          tracks[[self$funParams()$valueName]] <- tracks[[self$funParams()$valueName]][
            !names(tracks[[self$funParams()$valueName]]) %in% removeTrackIDs
          ]
        }
        
        # create data.table with properties
        tracks.DT <- tracks.measure.fun(
          tracks, celltrackR::speed,
          "live.cell.speed", steps.subtracks = 1,
          idcol = "cell_type", increment.cell.id = TRUE)
        
        # add further measurements
        tracks.DT[
          tracks.measure.fun(
            tracks, celltrackR::overallAngle, "live.cell.angle",
            steps.subtracks = 2, idcol = "cell_type",
            increment.cell.id = TRUE),
          on = .(cell_type, track_id, cell_id),
          live.cell.angle := .(live.cell.angle)]
        
        # push back to labels
        labels <- cciaObj$labelProps(valueName = self$funParams()$valueName)
        
        self$writeLog(self$funParams()$valueName)
        
        # get label obs
        label.obs <- as.data.table(
          labels$view_cols(list("label", "track_id", "cell_id"))$as_df(close = FALSE, include_x = FALSE))
        
        # check that there are no Nan cell id
        # these can occure when tracks were edited
        # if (any(is.na(label.obs$cell_id))) {
        label.obs[, cell_id := order(centroid_t), by = track_id]
        
        # get cell state information
        mergedDT <- tracks.DT[, c(
          "track_id", "cell_id", "live.cell.speed", "live.cell.angle"
        )][
          # https://stackoverflow.com/a/58508599
          # recalculate speed and angle
          # as.data.table(labels$values_obs())[, .SD, .SDcols = !c(
          #   "live.cell.speed", "live.cell.angle"
          # )], on = .(track_id, cell_id)
          # as.data.table(labels$values_obs()), on = .(track_id, cell_id)
          label.obs, on = .(track_id, cell_id)
          ]
        
        # set track id to nan where filtered
        mergedDT[track_id %in% removeTrackIDs, track_id := "NaN"]
        
        # add to labels
        labels$add_obs(
          as.list(mergedDT[, .(track_id, cell_id, live.cell.speed, live.cell.angle)])
        )
        
        # save
        labels$save()
        labels$close()
      }
      
      # update image information
      self$updateImageInfo()
      
      # DONE
      self$writeLog("Done")
      self$exitLog()
    }
  )
)

File: app/src/tasks/tracking/bayesian_tracking.jl
struct BayesianTracking <: CciaTask end

# Typed shape of what `_run_task(::BayesianTracking, …)` reads from `params`. Names mirror the
# btrack Bayesian tracker's inputs (accuracy / probToAssign / noise* / lambda* / theta* / *Thresh);
# defaults line up with the spec JSON so an omitted key parses identically to what the handler
# used to fall back to.
Base.@kwdef struct BayesianTrackingParams
    valueName::String            = VERSIONED_DEFAULT_VAL
    popsToTrack::String          = "NONE"
    trackSourceForce::Bool       = false
    maxSearchRadius::Int         = 20
    maxLost::Int                 = 3
    trackBranching::Bool         = false
    minTimepoints::Int           = 5
    accuracy::Float64            = 0.8
    probToAssign::Float64        = 0.8
    noiseInital::Int             = 300
    noiseProcessing::Int         = 100
    noiseMeasurements::Int       = 100
    distThresh::Float64          = 10.0
    timeThresh::Int              = 5
    segmentationMissRate::Float64 = 0.1
    lambdaLink::Int              = 5
    lambdaBranch::Int            = 50
    lambdaTime::Int              = 5
    lambdaDist::Float64          = 5.0
    thetaTime::Int               = 5
    thetaDist::Float64           = 5.0
end

function parse_bayesian_tracking_params(d::AbstractDict)::BayesianTrackingParams
    BayesianTrackingParams(;
        valueName            = string(get(d, "valueName", VERSIONED_DEFAULT_VAL)),
        popsToTrack          = string(get(d, "popsToTrack", "NONE")),
        trackSourceForce     = Bool(get(d, "trackSourceForce", false)),
        maxSearchRadius      = Int(get(d, "maxSearchRadius", 20)),
        maxLost              = Int(get(d, "maxLost", 3)),
        trackBranching       = Bool(get(d, "trackBranching", false)),
        minTimepoints        = Int(get(d, "minTimepoints", 5)),
        accuracy             = Float64(get(d, "accuracy", 0.8)),
        probToAssign         = Float64(get(d, "probToAssign", 0.8)),
        noiseInital          = Int(get(d, "noiseInital", 300)),
        noiseProcessing      = Int(get(d, "noiseProcessing", 100)),
        noiseMeasurements    = Int(get(d, "noiseMeasurements", 100)),
        distThresh           = Float64(get(d, "distThresh", 10.0)),
        timeThresh           = Int(get(d, "timeThresh", 5)),
        segmentationMissRate = Float64(get(d, "segmentationMissRate", 0.1)),
        lambdaLink           = Int(get(d, "lambdaLink", 5)),
        lambdaBranch         = Int(get(d, "lambdaBranch", 50)),
        lambdaTime           = Int(get(d, "lambdaTime", 5)),
        lambdaDist           = Float64(get(d, "lambdaDist", 5.0)),
        thetaTime            = Int(get(d, "thetaTime", 5)),
        thetaDist            = Float64(get(d, "thetaDist", 5.0)))
end

# Bayesian (btrack) cell tracking. Ports old-R-shiny bayesianTracking.R.
#
# Tracks either the whole segmentation or a gated flow population. Membership for the
# gated case is computed IN-PROCESS — Julia is the sole gate evaluator (docs/POPULATION.md),
# so we hand the label-ID list directly to Python (no CSV / no HTTP callback). btrack runs
# in the Python subprocess and writes the lineage columns (track_id, track_parent,
# track_root, track_state, track_generation, cell_id) back into the segmentation's
# labelProps/{valueName}.h5ad obs. No track measures / filters are computed here — gating
# on track properties is a later phase (docs/POPULATION.md).
function _run_task(task::BayesianTracking, img::CciaImage, params::Dict{String,Any};
                   on_log::Function      = line -> println(line),
                   on_progress::Function = (n, t) -> nothing,
                   on_process::Function  = _ -> nothing)

    p        = parse_bayesian_tracking_params(params)
    task_dir = img._dir
    props_path = img_label_props_path(img, p.valueName)
    if !isfile(props_path)
        on_log("[ERROR] No labelProps for valueName='$(p.valueName)': $props_path")
        return nothing
    end

    # Resolve gated-population membership in-process (Julia is the gate evaluator).
    label_ids = nothing
    # `trackSource` is the STABLE key `_write_back` uses to delete this pop's rows on a re-run —
    # the pop's UID (unchanged by rename/move) for a gated run, `WHOLE_SEG_TRACK_SOURCE` when tracking
    # the whole segmentation. See MULTI_POP_TRACKING_PLAN.md Decision 1.
    track_source = WHOLE_SEG_TRACK_SOURCE
    # Load the flow pop map up-front (even for a whole-seg run) so the P3 orphan sweep in
    # `_write_back` can see the current live UIDs and NaN any row still stamped by a deleted pop's
    # `track_source` — see MULTI_POP_TRACKING_ORPHANS_PLAN.md decision 3. No sidecar → no live pops,
    # every non-whole_seg row is an orphan; `nothing` there disables the sweep entirely on the
    # Python side, so a legacy tracking task that never emits the param still lands somewhere sane.
    m = try
        load_pop_map(img; value_name = p.valueName, pop_type = "flow")
    catch _
        nothing
    end
    # Shadow the params struct only inside the comprehension so `p` still refers to it below.
    live_track_sources = m === nothing ? nothing : [pop_uid(m, path) for path in pop_paths(m)]
    if p.popsToTrack != "NONE"
        m === nothing && (on_log("[ERROR] No gating sidecar for value_name='$(p.valueName)'"); return nothing)
        if !has_pop(m, p.popsToTrack)
            on_log("[ERROR] Population not found in gating/$(p.valueName).json: $(p.popsToTrack)")
            return nothing
        end
        recompute!(m, cols -> (label_props(img; value_name = p.valueName) |>
                               lp -> select_cols(lp, cols) |> as_df))
        label_ids = collect(Int, cells_in_pop(m, p.popsToTrack))
        track_source = pop_uid(m, p.popsToTrack)
        on_log("[INFO] Tracking $(length(label_ids)) cells from population '$(p.popsToTrack)' (uid=$track_source)")
        if isempty(label_ids)
            on_log("[ERROR] Population '$(p.popsToTrack)' is empty — nothing to track")
            return nothing
        end
    else
        on_log("[INFO] Tracking whole segmentation '$(p.valueName)'")
    end

    on_log("[INFO] Tracking labelProps: $props_path")

    # µm per pixel, skimage order [sz, sy, sx] — the same accessor every other spatial task uses
    # (cellNeighbours, the mesh tasks, track_measures). Tracking was the only one not calling it, so
    # the linking ran in pixels while the measures computed on its own output ran in µm.
    (pixel_res, _time_step) = img_physical_sizes(img)

    ok = run_py("tasks/tracking/bayesian_tracking_run.py",
        (; taskDir              = task_dir,
           physicalSizes        = pixel_res,
           valueName            = p.valueName,
           labelIds             = label_ids,                          # null = whole segmentation
           trackSource          = track_source,                       # pop UID or "whole_seg"
           # Override the P1 conflict detector: allow writing over labels currently owned by a
           # different pop's track_source. Only for the intentional pop→pop refinement idiom; the
           # whole-seg→pop case doesn't need it (whole_seg is treated as bypass in the detector).
           # Not exposed as a param widget yet — wired for future use.
           trackSourceForce     = p.trackSourceForce,
           # Live pop UIDs seen when this run launched. `nothing` (no sidecar loaded) disables the
           # P3 sweep — matches the legacy behaviour. See MULTI_POP_TRACKING_ORPHANS_PLAN decision 3.
           liveTrackSources     = live_track_sources,
           maxSearchRadius      = p.maxSearchRadius,
           maxLost              = p.maxLost,
           trackBranching       = p.trackBranching,
           minTimepoints        = p.minTimepoints,
           accuracy             = p.accuracy,
           probToAssign         = p.probToAssign,
           noiseInital          = p.noiseInital,
           noiseProcessing      = p.noiseProcessing,
           noiseMeasurements    = p.noiseMeasurements,
           distThresh           = p.distThresh,
           timeThresh           = p.timeThresh,
           segmentationMissRate = p.segmentationMissRate,
           lambdaLink           = p.lambdaLink,
           lambdaBranch         = p.lambdaBranch,
           lambdaTime           = p.lambdaTime,
           lambdaDist           = p.lambdaDist,
           thetaTime            = p.thetaTime,
           thetaDist            = p.thetaDist),
        task_run_dir(task_dir);
        on_log = on_log, on_progress = on_progress, on_process = on_process)
    ok || return nothing

    on_log("[INFO] Tracking complete.")

    # QC (advisory): bank the track count + mean track length from the track_id column btrack just
    # wrote back into the segmentation's obs. 0 tracks (btrack linked nothing) is the one unambiguous
    # problem → an advisory finding; the counts are always recorded as metrics for cohort stats.
    try
        cells = label_props(props_path) |> select_cols(["track_id"]) |> as_df
        tids  = "track_id" in names(cells) ? cells.track_id : Float64[]
        n_tracks, mean_len, n_tracked = track_count_metrics(tids)
        findings = n_tracks == 0 ?
            [qc_finding("warn", "tracking.no_tracks", "No tracks formed",
                "btrack linked no cells into tracks — check segmentation continuity and the tracking parameters, then re-run.")] :
            Dict{String,Any}[]
        write_qc(img, "tracking.bayesian_tracking", p.valueName, findings;
                 metrics = Dict{String,Any}("nTracks"         => n_tracks,
                                            "meanTrackLength" => round(mean_len, digits = 2),
                                            "nTrackedCells"   => n_tracked))
        on_log(n_tracks == 0 ? "[QC] no tracks formed — see the image's QC badge." :
               "[QC] $n_tracks track(s), mean length $(round(mean_len, digits = 1)) frames.")
    catch e
        on_log("[QC] could not compute tracking QC: $e")
    end

    Dict{String,Any}("valueName" => p.valueName)
end
